File: backend/app/schemas/voice.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

VoiceSessionStatus = Literal["pending", "active", "completed", "failed"]
VoiceTerminationReason = Literal[
    "session_end",
    "client_disconnect",
    "pipeline_error",
    "session_timeout",
]
VoiceEventType = Literal[
    "transcript",
    "tool_call",
    "tool_result",
    "agent_text",
    "error",
    "session_end",
]


class VoiceSessionCreateRequest(BaseModel):
    """Body of POST /api/v1/voice/sessions. Tunable knobs for the demo session."""

    system_prompt: str = Field(min_length=1, max_length=4000)
    cartesia_voice_id: str = Field(min_length=1)
    openai_model_id: str = "gpt-4o-mini"

    model_config = {"extra": "forbid"}


class VoiceSessionCreateResponse(BaseModel):
    """Response of POST /api/v1/voice/sessions."""

    session_id: str
    websocket_path: str
    artifact_path: str
    status: VoiceSessionStatus = "pending"

    model_config = {"extra": "forbid"}


class VoiceSessionEventEnvelope(BaseModel):
    """Typed envelope for non-audio control frames sent to the client over the WebSocket.

    Audio frames are raw PCM and not wrapped in JSON.
    """

    type: VoiceEventType
    payload: dict[str, object]

    model_config = {"extra": "forbid"}

File: frontend/src/types/voice.ts
export type VoiceSessionStatus = "idle" | "connecting" | "active" | "failed" | "finalized";

export type VoiceEventType =
  | "transcript"
  | "tool_call"
  | "tool_result"
  | "agent_text"
  | "error"
  | "session_end";

export type VoiceTerminationReason =
  | "session_end"
  | "client_disconnect"
  | "pipeline_error"
  | "session_timeout";

export interface VoiceSessionCreateRequest {
  readonly system_prompt: string;
  readonly cartesia_voice_id: string;
  readonly openai_model_id?: string;
}

export interface VoiceSessionCreateResponse {
  readonly session_id: string;
  readonly websocket_path: string;
  readonly artifact_path: string;
  readonly status: "pending" | "active" | "completed" | "failed";
}

export interface VoiceSessionEventEnvelope {
  readonly type: VoiceEventType;
  readonly payload: Record<string, unknown>;
}

export interface TranscriptEntry {
  readonly index: number;
  readonly role: "user" | "assistant";
  readonly text: string;
  readonly startedAt: string;
  readonly endedAt: string | null;
  readonly isInterim: boolean;
}

export interface ToolTraceEntry {
  readonly index: number;
  readonly toolCallId: string;
  readonly name: string;
  readonly arguments: Record<string, unknown>;
  readonly result: Record<string, unknown> | null;
  readonly startedAt: string;
  readonly endedAt: string | null;
  readonly durationMs: number | null;
}
